include/seqan3/search/detail/unidirectional_search_algorithm.hpp

```diff
@@ -37,17 +41,127 @@ enum class error_type : uint8_t
+    index_t const * index_ptr{nullptr};
+
+    //!\brief A function object that stores the on-hit-delegate to be executed whenever a hit in the index is found.
+    std::function<void(typename index_t::cursor_type const &)> delegate;
+
+    // forward declaration
+    template <bool abort_on_hit, typename query_t>
+    bool search_trivial(typename index_t::cursor_type cur,
+                        query_t & query,
+                        typename index_t::cursor_type::size_type const query_pos,
+                        search_param const error_left,
+                        error_type const prev_error);
+
+    //!\brief Befriend seqan3::detail::test_accessor to grant access to layout.
+    friend struct ::seqan3::detail::test_accessor;
+public:
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    unidirectional_search_algorithm() = default; //!< Defaulted.
+    unidirectional_search_algorithm(unidirectional_search_algorithm const &) = default; //!< Defaulted.
+    unidirectional_search_algorithm(unidirectional_search_algorithm &&) = default; //!< Defaulted.
+    unidirectional_search_algorithm & operator=(unidirectional_search_algorithm const &) = default; //!< Defaulted.
+    unidirectional_search_algorithm & operator=(unidirectional_search_algorithm &&) = default; //!< Defaulted.
+    ~unidirectional_search_algorithm() = default; //!< Defaulted.
+
+    /*!\brief Constructs from a configuration object and an index.
+     * \tparam configuration_t The search configuration type.
+     * \tparam index_t The type of index; must model seqan3::fm_index_specialisation.
+     * \param[in] cfg The configuration object that guides the search algorithm.
+     * \param[in] index The index used in the algorithm.
+     */
+    unidirectional_search_algorithm(configuration_t const & cfg, index_t const & index) : policies_t{}...
+    {
+        config = cfg;
+        index_ptr = &index;
+    }
+    //!\}
+
+    /*!\brief Searches a query sequence in an FM index using trivial backtracking.
+     * \tparam query_t       Must model std::ranges::input_range over the index's alphabet.
+     * \param[in] query      Query sequence to be searched in the index.
+     *
+     * ### Complexity
+     *
+     * \f$O(|query|^e)\f$ where \f$e\f$ is the maximum number of errors.
+     */
+    template <typename query_t>
+    auto operator()(query_t && query) noexcept
+    {
+        auto max_error = this->get_search_param(config, query); // see policy_max_error
+
+        // construct internal delegate for collecting hits for later filtering (if necessary)
+        std::vector<typename index_t::cursor_type> internal_hits;
+        delegate = [&internal_hits] (auto const & it)
+        {
+            internal_hits.push_back(it);
+        };
+
+        // choose mode
+        if constexpr (traits_t::search_best_hits || traits_t::search_all_best_hits || traits_t::search_strata_hits)
```

Comment:
As said above. Making this into a policy would require to make all policies into CRTP base classes and I tried to avoid this. Can this be part of a later refactoring?

Reply:
I doubt the CRTP class, but lets defer the policy.

I still would want the spilt, because this function is packed.